vscripts/confogl.nut

```diff
@@ -2,65 +2,308 @@
+    local nextPlugin = pluginScopes[idx];
+    if ("AllowBash" in pluginScopes[idx]) {
+	  local fn = pluginScopes[idx].AllowBash;
+	  if ( typeof (fn) == "function" ) {
+	    local res = fn(basher, bashee)
+	    // If any plugin returns ALLOW_BASH_NONE, we return ALLOW_BASH_NONE.
+	    if(res == ALLOW_BASH_NONE) {
+		  return ALLOW_BASH_NONE;
+	    }
+	    // If any plugin returns ALLOW_BASH_PUSHONLY, we return ALLOW_BASH_PUSHONLY.
+	    if(res == ALLOW_BASH_PUSHONLY) {
+		  return ALLOW_BASH_PUSHONLY;
+	    }
+	  }
+    }
+  }
+  return retVal;
+}
+// bool AllowTakeDamage(table damageTable): Adding a function with this name in the script causes C++ to call it on all damage events.
+// The damageTable is actually defined in scriptedmode.nut and filled in as appropriate before each call.
+function AllowTakeDamage(damageTable) {
+  Msg(format("AllowTakeDamage(%s)\n", damageTable.tostring()));
+  // Default to allow the damage.
+  local retVal = true;
+  for ( local idx = pluginScopes.len() - 1; idx >= 0; --idx ) {
+    local nextPlugin = pluginScopes[idx];
+    if ("AllowTakeDamage" in pluginScopes[idx]) {
+	  local fn = pluginScopes[idx].AllowTakeDamage;
+	  if ( typeof (fn) == "function" ) {
+	    local res = fn(damageTable)
+	    // If any plugin returns false, we return false.
+	    if(res == false) {
+		  return false;
+	    }
+	  }
+    }
+  }
+  return retVal;
+}
+// bool BotQuery(int queryflag, handle entity, bool defaultvalue): Hook to control survivor bot behavior.
+// Only one known flag is supported.
+// BOT_QUERY_NOTARGET - Fired when a bot wants to bash a prop. Returning false disallows the bashing.
+function BotQuery(queryflag, entity, defaultvalue) {
+  Msg(format("BotQuery(%d, %s, %s)\n", queryflag, entity.tostring(), defaultvalue));
+  // Default to allow the bash.
+  local retVal = true;
+  for ( local idx = pluginScopes.len() - 1; idx >= 0; --idx ) {
+    local nextPlugin = pluginScopes[idx];
+    if ("BotQuery" in pluginScopes[idx]) {
+	  local fn = pluginScopes[idx].BotQuery;
+	  if ( typeof (fn) == "function" ) {
+	    local res = fn(queryflag, entity, defaultvalue)
+	    // If any plugin returns false, we return false.
+	    if(res == false) {
+		  return false;
+	    }
+	  }
+    }
+  }
+  return retVal;
+}
+// bool CanPickupObject(handle object): Hook function for deciding whether a given prop should be pickupable (HL2 style).
+// Returning true makes the players able to pick up the object.
+// Only functional for server side props. If this is defined it must return true in order the PickupObject() function to work. 
+function CanPickupObject(object) {
+  Msg(format("CanPickupObject(%s)\n", object.tostring()));
+  // Default to disallow the pickup.
+  local retVal = false;
+  for ( local idx = pluginScopes.len() - 1; idx >= 0; --idx ) {
+    local nextPlugin = pluginScopes[idx];
+    if ("CanPickupObject" in pluginScopes[idx]) {
+	  local fn = pluginScopes[idx].CanPickupObject;
+	  if ( typeof (fn) == "function" ) {
```

Comment:
The above 5 lines look to be the same in every function. Let's refactor them to a helper function:

There's a few options here, but something like:
```squirrel
function forEachPluginHook(hookName, cb) {
  for ( local idx = pluginScopes.len() - 1; idx >= 0; --idx ) {
    local nextPlugin = pluginScopes[idx];
    if (hookName in pluginScopes[idx]) {
      local fn = pluginScopes[idx][HookName];
      if ( typeof (fn) == "function" ) {
        cb(fn);
      }
    }
  }
}  
```

Which would then give usage like:

```squirrel

function CanPickupObject(object) {
  Msg(format("CanPickupObject(%s)\n", object.tostring()));
  // Default to disallow the pickup.
  local retVal = false;
  forEachPluginHook("CanPIckupObject", [this, object], function (hookFn) {
    local res = hookFn(object);
    if (res == true) {
      return true;
    }
  });
  return retVal;
}
```

Reply:
Implemented: https://github.com/Icy-Inferno/confogl-mutation/commit/f38354693dcdf832d2afd46c73a08962ff786002